pkg/execution/pauses/iterator.go

```diff
@@ -0,0 +1,244 @@
+package pauses
+
+import (
+	"context"
+	"slices"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/inngest/inngest/pkg/execution/state"
+	"github.com/oklog/ulid/v2"
+)
+
+const (
+	// DefaultConcurrentBlockFetches represents the number of block fetches we run
+	// concurrently.  Note that this will consume _at least_ N * DefaultPausesPerBlock
+	// KB of memory.
+	//
+	// With a default of 20, this is approx 500MB of memory.
+	DefaultConcurrentBlockFetches = 20
+)
+
+func newDualIter(idx Index, bufferedIter state.PauseIterator, rdr BlockReader, blockIDs []ulid.ULID) *dualIter {
+	// NOTE: This is just an estimate, as pauses may have been compacted / had deletions.
+	count := bufferedIter.Count() + (len(blockIDs) * DefaultPausesPerBlock)
+
+	return &dualIter{
+		idx:             idx,
+		count:           count,
+		bufferIter:      bufferedIter,
+		blockReader:     rdr,
+		unfetchedBlocks: blockIDs,
+		l:               &sync.Mutex{},
+	}
+}
+
+// dualIter represents an iterator that reads from blocks as well as buffers,
+// downloading blocks in parallel to maximize throughput.
+type dualIter struct {
+	idx Index
+
+	// index is the current iterator index.
+	index int64
+
+	// count is an esitmate of the max pauses in the iterator.
+	count int
+
+	// usingBuffer indicates whether we're using the buffer for the next
+	// .Val() call.  This is set to true to begin with, while the buffer iter
+	// is being used.
+	usingBuffer bool
+
+	// bufferIter is the buffered iterator.
+	bufferIter state.PauseIterator
+
+	// blockReader is the block reader to fetch metadata and blocks.
+	blockReader BlockReader
+
+	// unfetchedBlocks represents the set of blocks that haven't been fetched
+	// from the backing store yet.
+	unfetchedBlocks []ulid.ULID
+
+	// inflightBlocks represents blocks that are currently being fetched from
+	// the backing store.
+	inflightBlocks []ulid.ULID
+
+	// pauses represents the current pauses that have been fetched from downloaded
+	// blocks.
+	pauses []*state.Pause
+
+	err error
+
+	// l represents a lock held when mutating block slices or pauses.
+	l *sync.Mutex
```

Comment:
can this be a rw lock or does it need to be a mutex?

I see a couple of cases where it's more reading than writing to it.

Reply:
AFAIK, there aren't concurrent reads so it doesn't matter.

Reply:
not necessary if there are no benefits to it